pallets/did/src/lib.rs

```diff
@@ -28,91 +30,408 @@ mod tests;
+use sp_core::{ed25519, sr25519};
+use sp_runtime::traits::Verify;
+use sp_std::{collections::btree_set::BTreeSet, fmt::Debug, prelude::Clone, vec::Vec};
+
+pub use pallet::*;
+
+/// Reference to a payload of data of variable size.
+pub type Payload = [u8];
+
+/// Type for an encoded URL.
+pub type UrlEncoding = Vec<u8>;
 
-/// The DID trait
-pub trait Config: frame_system::Config {
-	/// DID specific event type
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+/// Trait representing a public key under the control of a DID subject.
+pub trait DIDPublicKey {
+	/// Returns the key method description as in the [DID specification](https://w3c.github.io/did-spec-registries/#verification-method-types).
+	fn get_did_key_description(&self) -> &'static str;
+}
 
-	/// Weight information for extrinsics in this pallet.
-	type WeightInfo: WeightInfo;
+/// An enum describing the different verification methods a verification key can
+/// fulfil, according to the [DID specification](https://w3c.github.io/did-spec-registries/#verification-relationships).
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq)]
+pub enum DIDVerificationKeyType {
+	Authentication,
+	CapabilityDelegation,
+	// Not used for now, but added for potential future use.
+	CapabilityInvocation,
+	AssertionMethod,
+}
 
-	/// Public signing key type for DIDs
-	type PublicSigningKey: Parameter + Member + Codec + Default;
+/// An enum describing the different verification methods an encryption key can
+/// fulfil, according to the [DID specification](https://w3c.github.io/did-spec-registries/#verification-relationships).
+#[derive(Clone, Debug, Decode, Encode, PartialEq)]
+pub enum DIDEncryptionKeyType {
+	KeyAgreement,
+}
 
-	/// Public boxing key type for DIDs
-	type PublicBoxKey: Parameter + Member + Codec + Default;
+/// Enum representing the types of verification keys a DID can control.
+#[derive(Clone, Decode, Debug, Encode, Eq, Ord, PartialEq, PartialOrd)]
+pub enum PublicVerificationKey {
+	/// An Ed25519 public key.
+	Ed25519(ed25519::Public),
+	/// A Sr25519 public key.
+	Sr25519(sr25519::Public),
 }
 
-decl_event!(
-	/// Events for DIDs
-	pub enum Event<T> where <T as frame_system::Config>::AccountId {
-		/// A did has been created
-		DidCreated(AccountId),
-		/// A did has been removed
-		DidRemoved(AccountId),
+impl From<ed25519::Public> for PublicVerificationKey {
+	fn from(key: ed25519::Public) -> Self {
+		PublicVerificationKey::Ed25519(key)
 	}
-);
+}
 
-decl_module! {
-	/// The DID runtime module
-	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+impl From<sr25519::Public> for PublicVerificationKey {
+	fn from(key: sr25519::Public) -> Self {
+		PublicVerificationKey::Sr25519(key)
+	}
+}
 
-		/// Deposit events
-		fn deposit_event() = default;
+impl PublicVerificationKey {
+	/// Given a payload and a signature, the specific public verification key
+	/// will return either a [SignatureError](SignatureError) if the signature
```

Comment:
Does this link `[SignatureError](SignatureError)` work for you/some plugin? It does not for me or on Github. Would suggest do drop it then

```suggestion
	/// will return either a `SignatureError` if the signature
```